### PR TITLE
feat: implement NL parser for command execution

### PR DESCRIPTION
**Pull request by pavanimanchala53**
## Related Issue
Closes cortexlinux/cortex-distro#25
## Summary
This PR improves the natural-language install flow by standardizing prompt
construction and ensuring deterministic command generation without hardcoded
package mappings.

## What changed
- Normalized prompt handling between CLI and CommandInterpreter
- Added unit tests validating NL → command generation for ML-style requests
- Added concise documentation for the NL installer feature
- Removed unreachable and duplicated prompt logic flagged by static analysis

## Testing
- Added `tests/test_nl_installer.py`
- Verified command generation without execution, sudo, or network access

- All tests pass via `pytest`

## Notes
- No hardcoded install logic or fixed package lists were introduced
- Execution remains opt-in via `--execute`

## AI Used
- ChatGPT
- Github Copilot

working video

with --execute flag

## Summary by CodeRabbit

* **New Features**
  * Public intent-extraction API with multi‑provider support, domain and install‑mode awareness, and provider-specific intent handling.
  * Enhanced interactive install flow: stdin-aware prompts, confidence-driven clarify/confirm/auto-approve, editable plans, and generated suggestions.

* **Improvements**
  * Domain/context included in prompts; more robust parsing of varied LLM outputs; improved fallbacks and empty-response handling.

* **Documentation**
  * Added Natural Language Installer guide (usage, dry‑run vs execute).

* **Tests**
  * New NL parser test suite covering diverse install and intent scenarios.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by pavanimanchala53** [comment]
hi @Anshgrover23 

here is the working video
can you please review this ?

https://github.com/user-attachments/assets/7f147835-6c35-4b6a-9480-c92ba93e8f12

File: cortex/llm/interpreter.py [hunk 1]
```diff
@@ -173,6 +258,50 @@ def _call_openai(self, user_input: str) -> list[str]:
...
+            messages=[
+                {"role": "system", "content": self._get_intent_prompt()},
+                {"role": "user", "content": user_input},
+            ],
+            temperature=0.2,
+            max_tokens=300,
+        )
+
+        content = response.choices[0].message.content.strip()
+        return json.loads(content)
```
[the remark this answers is not in the document]
Comment (Anshgrover23) [reply cortex/llm/interpreter.py]: @pavanimanchala53 Address this one as well.

File: cortex/cli.py [hunk 2]
```diff
@@ -609,6 +652,55 @@ def install(
             for i, cmd in enumerate(commands, 1):
                 print(f"  {i}. {cmd}")
 
+            # ---------- User confirmation ----------
+            # ---------- User confirmation ----------
+            if execute:
+                if not _is_interactive():
+                    # Non-interactive mode (pytest / CI) → auto-approve
```
[the remark this answers is not in the document]
Comment (pavanimanchala53) [reply cortex/cli.py]: reverted the non-interactive execution restriction and restored the original interactive confirmation flow(y/e/n with enter defaulting to yes) this aligns with existing CLI behavior and test expectations.

File: cortex/llm/interpreter.py [hunk 3]
```diff
@@ -385,3 +527,50 @@ def parse_with_context(
 
         enriched_input = user_input + context
         return self.parse(enriched_input, validate=validate)
+
+    def _estimate_confidence(self, user_input: str, domain: str) -> float:
```
[the remark this answers is not in the document]
Comment (pavanimanchala53) [reply cortex/llm/interpreter.py]: addressed now

File: cortex/cli.py [hunk 4]
```diff
@@ -622,6 +706,56 @@ def install(
...
+
+                    confirm = input("\nExecute edited commands? [y/n]: ").strip().lower()
+                    if confirm != "y":
+                        print("❌ Installation cancelled.")
+                        return 0
+
+                elif choice != "y":
+                    print("❌ Invalid choice. Cancelling.")
+                    return 1
+            # -------------------------------------
```
[the remark this answers is not in the document]
Comment (pavanimanchala53) [reply cortex/cli.py]: when user edits the commands already generated validation is not necessary

File: cortex/cli.py [hunk 5]
```diff
@@ -32,11 +32,55 @@
...
+        Returns True if the request is too underspecified to safely proceed.
+        """
+        if not intent:
+            return True
+
+        domain = intent.get("domain", "unknown")
+        if domain == "unknown":
+            return True
+
+        return False
```
[the remark this answers is not in the document]
Comment (pavanimanchala53) [reply cortex/cli.py]: we intentionally do not recheck the ambiguous field in cli.py . Ambiguity is already handled in interpreter layer where the LLM is explicitly prompted to determine and return ambiguous as part of intent content.

File: cortex/cli.py [hunk 6]
```diff
@@ -32,11 +32,55 @@
...
+        Combine optional stdin context with user prompt.
+        """
+        stdin_data = getattr(self, "stdin_data", None)
+        if stdin_data:
+            return (
+                "Context (from stdin):\n" f"{stdin_data}\n\n" "User instruction:\n" f"{user_prompt}"
+            )
+        return user_prompt
+
+    def _is_ambiguous_request(self, user_input: str, intent: dict | None) -> bool:
```
Comment (Anshgrover23) [inline cortex/cli.py]: Ambiguity detection works, but the flow stops execution instead of prompting for clarification and continuing. This doesn’t fully meet “handle ambiguity gracefully” for demos.

File: tests/test_nl_parser_cases.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+import pytest
+
+from cortex.llm.interpreter import CommandInterpreter
+
+
+@pytest.fixture
+def fake_interpreter(monkeypatch):
+    monkeypatch.setenv(
+        "CORTEX_FAKE_COMMANDS",
+        '{"commands": ["echo install step 1", "echo install step 2"]}',
```
Comment (Anshgrover23) [inline tests/test_nl_parser_cases.py]: Fake provider bypasses intent + ambiguity logic entirely. Why it is hiding edge cases ?

**Review by Anshgrover23 — CHANGES_REQUESTED** [review]
@pavanimanchala53 Some issue requirements are still incomplete:
- Intent reasoning is computed but never displayed. The issue explicitly asks to show “I understood you want…”, which is currently missing.
- Also, the clarification prompt is static and does not suggest alternatives dynamically, which is part of the issue requirements.

**Review by Anshgrover23 — CHANGES_REQUESTED** [review]
@pavanimanchala53 CI checks are failing.

File: cortex/cli.py [hunk 8]
```diff
@@ -675,15 +908,95 @@ def install(
...
+
+            if confidence_level == "low":
+                # Low confidence: ask user to clarify what they want
+                if _is_interactive():
+                    clarification_msg = self._generate_clarification_request(
+                        interpreter, software, intent
+                    )
+                    print(f"\n{clarification_msg}\n")
+                    clarified = input("What would you like to install? ").strip()
+                    if clarified:
```
[the remark this answers is not in the document]
Comment (Anshgrover23) [reply cortex/cli.py]: @pavanimanchala53 Address this one.

File: cortex/cli.py [hunk 9]
```diff
@@ -36,11 +38,237 @@
...
+            elif interpreter.provider.name == "claude":
+                response = interpreter.client.messages.create(
+                    model=interpreter.model,
+                    max_tokens=max_tokens,
+                    temperature=temperature,
+                    system=system_message,
+                    messages=[{"role": "user", "content": prompt}],
+                )
+                return response.content[0].text.strip()
+            elif interpreter.provider.name == "ollama":
```
Comment (Anshgrover23) [inline cortex/cli.py]: If interpreter.provider is not OLLAMA during `__init__`, ollama_url won't exist. However, this path uses it. ,it is case of fragile coupling.

File: cortex/cli.py [hunk 10]
```diff
@@ -36,11 +39,220 @@
 sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))
 
 
+def _is_interactive():
+    return sys.stdin.isatty()
```
[the remark this answers is not in the document]
Comment (Anshgrover23) [reply cortex/cli.py]: @pavanimanchala53 Address this one.

File: cortex/cli.py [hunk 11]
```diff
@@ -32,11 +35,172 @@
...
+
```
[the remark this answers is not in the document]
Comment (Anshgrover23) [reply cortex/cli.py]: same do this.

**Review by Anshgrover23 — CHANGES_REQUESTED** [review]
@pavanimanchala53 Kindly resolve conflicts.
